keyError: target_km_walked

I'm new to Python but was able to get the environment all setup and running.  The bot starts up and logs in fine but shortly after trying to walk to the first pokestop I get an error:

2016-08-03 00:20:36,632 [   pokebot] [ERROR] Main loop has an ERROR, restarting 'target_km_walked'
Traceback (most recent call last):
  File "pokebot.py", line 99, in main
    api.main_loop()
  File "C:\Users\lennon68\pokeBot\PoGoBoWo\pgoapi\pgoapi.py", line 601, in main_loop
    self.heartbeat()
  File "C:\Users\lennon68\pokeBot\PoGoBoWo\pgoapi\pgoapi.py", line 271, in heartbeat
    self.log.info("\n\nList of Pokemon:\n" + get_inventory_data(res, self.pokemon_names) + "\nTotal Pokemon count: " + str(get_pokemon_num(res)) + "\nEgg Hatching status: " + get_incubators_stat(res) + "\n")
  File "C:\Users\lennon68\pokeBot\PoGoBoWo\pgoapi\poke_utils.py", line 56, in get_incubators_stat
    km_walked), inventory_items_incubator_list)))
  File "C:\Users\lennon68\pokeBot\PoGoBoWo\pgoapi\poke_utils.py", line 55, in <lambda>
    x['target_km_walked'],
KeyError: 'target_km_walked'
2016-08-03 00:21:06,812 [auth_google] [ INFO] Google login for:

It is hitting some pokestops (although it's saying it has less than 1 poke ball so I don't know how that's working...).  Config file attached.

[config_clean.txt](https://github.com/BoBeR182/PoGoBoWo/files/398703/config_clean.txt)

--- a/config.json.example
+++ b/config.json.example
@@ -14,6 +14,7 @@
         "RETURN_START_INTERVAL": 20,
         "MAX_BALL_TYPE": 2,
         "SLOW_BUT_STEALTH": false,
+        "AUTO_HATCHING": false,
         "MIN_ITEM_COUNTS": {
             "ITEM_POKE_BALL": 15,
             "ITEM_GREAT_BALL": 25,
@@ -42,6 +43,7 @@
         "RETURN_START_INTERVAL": 20,
         "MAX_BALL_TYPE": 2,
         "SLOW_BUT_STEALTH": false,
+        "AUTO_HATCHING": false,
         "MIN_ITEM_COUNTS": {
             "ITEM_POKE_BALL": 15,
             "ITEM_GREAT_BALL": 25,

--- a/pgoapi/pgoapi.py
+++ b/pgoapi/pgoapi.py
@@ -20,7 +20,8 @@ import random
 import json
 from pgoapi.location import distance_in_meters, get_increments, get_neighbors, get_route, filtered_forts
 # import pgoapi.protos.POGOProtos.Enums_pb2 as RpcEnum
-from pgoapi.poke_utils import pokemon_iv_percentage, get_inventory_data, get_pokemon_num, get_incubators_stat
+from pgoapi.poke_utils import pokemon_iv_percentage, get_inventory_data, get_pokemon_num, get_incubators_stat, incubators_stat_str, \
+        get_eggs_stat
 from time import sleep
 from collections import defaultdict
 import os.path
@@ -156,6 +157,7 @@ class PGoApi:
         self.DUPLICATE_CP_FORGIVENESS = config.get("DUPLICATE_CP_FORGIVENESS", 0)
         self.MAX_BALL_TYPE = config.get("MAX_BALL_TYPE", 0)
         self.SLOW_BUT_STEALTH = config.get("SLOW_BUT_STEALTH", 0)
+        self.AUTO_HATCHING = config.get("AUTO_HATCHING", False)
         self._req_method_list = []
         self._heartbeat_number = 0
         self.pokemon_names = pokemon_names
@@ -268,7 +270,7 @@ class PGoApi:
             # new inventory data has just been saved, clearing evolved pokemons list
             self.evolved_pokemon_ids = []
             # create string with pokemon list, add users info and print everything
-            self.log.info("\n\nList of Pokemon:\n" + get_inventory_data(res, self.pokemon_names) + "\nTotal Pokemon count: " + str(get_pokemon_num(res)) + "\nEgg Hatching status: " + get_incubators_stat(res) + "\n")
+            self.log.info("\n\nList of Pokemon:\n" + get_inventory_data(res, self.pokemon_names) + "\nTotal Pokemon count: " + str(get_pokemon_num(res)) + "\nEgg Hatching status: " + incubators_stat_str(res) + "\n")
             self.log.info("\n\n Username: %s, Lvl: %s, XP: %s/%s \n Currencies: %s \n", player_data.get('username', 'NA'), player_stats.get('level', 'NA'), player_stats.get('experience', 'NA'), player_stats.get('next_level_xp', 'NA'), currency_data)
             self.log.debug(self.cleanup_inventory(res['responses']['GET_INVENTORY']['inventory_delta']['inventory_items']))
         self._heartbeat_number += 1
@@ -293,6 +295,10 @@ class PGoApi:
                             sleep(3 * random.random() + 1) # If you want to make it faster, delete this line... would not recommend though
                         else:
                             sleep(1)
+        if self.AUTO_HATCHING:
+            hatching_eggs_count = self.attempt_hatch_eggs()
+            if hatching_eggs_count > 0:
+                self.log.info("Start hatching %d eggs", hatching_eggs_count)
 
     # this is in charge of spinning a pokestop
     def spin_near_fort(self):
@@ -594,6 +600,39 @@ class PGoApi:
             sleep(3 * random.random() + 10)
 
         return True
+
+    def attempt_hatch_eggs(self, res=None):
+        self.get_hatched_eggs().call()
+        if not res:
+            res = self.get_inventory().call()
+        hatching_incubator_list, empty_incubator_list = get_incubators_stat(res)
+        hatching_eggs, immature_eggs = get_eggs_stat(res)
+        hatching_eggs_count = 0
+        for immature_egg in immature_eggs:
+            egg_id = immature_egg['pokemon_data']['id']
+            if len(empty_incubator_list) > 0:
+                # Always use first incubator.
+                incubator_index = 0
+                incubator_id = empty_incubator_list[incubator_index]['id']
+                uses_remaining = empty_incubator_list[incubator_index].get('uses_remaining', 0) - 1
+                if self.hatch_egg(incubator_id, egg_id):
+                    hatching_eggs_count += 1
+                    # Update incubator manually to save api call().
+                    empty_incubator_list[incubator_index]['uses_remaining'] = uses_remaining
+                    if uses_remaining <= 0:
+                        del(empty_incubator_list[incubator_index])
+        return hatching_eggs_count
+
+
+    def hatch_egg(self, incubator_id, egg_id):
+        response = self.use_item_egg_incubator(item_id = incubator_id, pokemon_id = egg_id).call()
+        result = response.get('responses', {}).get('USE_ITEM_EGG_INCUBATOR', {})
+        if len(result) == 0:
+            return False
+        if "result" in result:
+            self.log.debug("Result: %d", result['result'])
+        return result.get('result', 0) == 1
+
 
     def main_loop(self):
         while True:

--- a/pgoapi/pgoapi.py
+++ b/pgoapi/pgoapi.py
@@ -21,7 +21,7 @@ import json
 from pgoapi.location import distance_in_meters, get_increments, get_neighbors, get_route, filtered_forts
 # import pgoapi.protos.POGOProtos.Enums_pb2 as RpcEnum
 from pgoapi.poke_utils import pokemon_iv_percentage, get_inventory_data, get_pokemon_num, get_incubators_stat, incubators_stat_str, \
-        get_eggs_stat
+    get_eggs_stat
 from time import sleep
 from collections import defaultdict
 import os.path
@@ -623,16 +623,14 @@ class PGoApi:
                         del(empty_incubator_list[incubator_index])
         return hatching_eggs_count
 
-
     def hatch_egg(self, incubator_id, egg_id):
-        response = self.use_item_egg_incubator(item_id = incubator_id, pokemon_id = egg_id).call()
+        response = self.use_item_egg_incubator(item_id=incubator_id, pokemon_id=egg_id).call()
         result = response.get('responses', {}).get('USE_ITEM_EGG_INCUBATOR', {})
         if len(result) == 0:
             return False
         if "result" in result:
             self.log.debug("Result: %d", result['result'])
         return result.get('result', 0) == 1
-
 
     def main_loop(self):
         while True:

--- a/pgoapi/poke_utils.py
+++ b/pgoapi/poke_utils.py
@@ -46,10 +46,10 @@ def get_incubators_stat(res):
     inventory_items_incubators = map(lambda x: x.get('egg_incubators', {}).get('egg_incubator', {}), inventory_items_dict_list)
     inventory_items_incubator_list = reduce(lambda x, y: x + y, filter(lambda x: len(x) > 0, inventory_items_incubators))
     player_stats = get_player_stats(res)
-    km_walked = player_stats.get('km_walked', 'None')
+    km_walked = player_stats.get('km_walked', -1)
     if inventory_items_incubator_list:
         return (os.linesep.join(map(lambda x: "Incubator {0:.2f} km, walked {1:.2f} km".format(
-            x['target_km_walked'],
+            x.get('target_km_walked', 0),
             km_walked), inventory_items_incubator_list)))
     else:
         return 'No incubators'

--- a/pgoapi/poke_utils.py
+++ b/pgoapi/poke_utils.py
@@ -41,21 +41,48 @@ def get_inventory_data(res, poke_names):
     return pk_list
 
 
+def get_player_stats(res):
+    inventory_items_dict_list = extract_items_dict(res)
+    player_stats = filter(lambda x: len(x) > 0, map(lambda x: x.get('player_stats', {}), inventory_items_dict_list))
+    if len(player_stats) > 0:
+        return player_stats[0]
+    else:
+        return {}
+
+
+def get_eggs_stat(res):
+    inventory_items_dict_list = extract_items_dict(res)
+    all_eggs = filter(lambda x: 'pokemon_data' in x and 'is_egg' in x['pokemon_data'], inventory_items_dict_list)
+    hatching_eggs = filter(lambda x: 'id' in x['pokemon_data'] and 'egg_incubator_id' in x['pokemon_data'], all_eggs)
+    immature_eggs = filter(lambda x: 'id' in x['pokemon_data'] and 'egg_incubator_id' not in x['pokemon_data'], all_eggs)
+    return hatching_eggs, immature_eggs
+
+
 def get_incubators_stat(res):
     inventory_items_dict_list = extract_items_dict(res)
     inventory_items_incubators = map(lambda x: x.get('egg_incubators', {}).get('egg_incubator', {}), inventory_items_dict_list)
     inventory_items_incubator_list = reduce(lambda x, y: x + y, filter(lambda x: len(x) > 0, inventory_items_incubators))
-    player_stats = get_player_stats(res)
-    km_walked = player_stats.get('km_walked', -1)
-    if inventory_items_incubator_list:
-        return (os.linesep.join(map(lambda x: "Incubator {0:.2f} km, walked {1:.2f} km".format(
-            x.get('target_km_walked', 0),
-            km_walked), inventory_items_incubator_list)))
+    inventory_items_hatching_incubator_list = filter(lambda x: 'id' in x and 'target_km_walked' in x, inventory_items_incubator_list)
+    inventory_items_empty_incubator_list = filter(lambda x: 'id' in x and 'target_km_walked' not in x, inventory_items_incubator_list)
+    return inventory_items_hatching_incubator_list, inventory_items_empty_incubator_list
+
+
+def incubators_stat_str(res):
+    hatching_incubator_list, empty_incubator_list = get_incubators_stat(res)
+    incubators_stat = ''
+    if len(hatching_incubator_list) > 0:
+        km_walked = get_player_stats(res).get('km_walked', -1)
+        incubators_stat += "\n" + (os.linesep.join(map(lambda x: "Incubator {0:.2f} km, walked {1:.2f} km {2}".format(
+            x['target_km_walked'],
+            km_walked,
+            '(Done)' if km_walked >= x['target_km_walked'] else '',
+        ), hatching_incubator_list)))
     else:
-        return 'No incubators'
-    if inventory_items_incubator_list:
-        return (os.linesep.join(map(lambda x: "Incubator {0:.2f} km, walked {1:.2f} km".format(
-            x['egg_incubator']['target_km_walked'],
-            x['egg_incubator']['start_km_walked']), inventory_items_incubator_list)))
+        incubators_stat += '\nNo hatching incubators'
+
+    if len(empty_incubator_list) > 0:
+        incubators_stat += "\n" + (os.linesep.join(map(lambda x: "Incubator id: {0}, uses_remaining: {1} time(s)".format(
+            x['id'], x.get('uses_remaining', 'inf')), empty_incubator_list)))
     else:
-        return 'No incubators'
+        incubators_stat += '\nAll incubators are using'
+    return incubators_stat

--- a/pgoapi/poke_utils.py
+++ b/pgoapi/poke_utils.py
@@ -6,18 +6,20 @@ def pokemon_iv_percentage(pokemon):
         'individual_defense', 0) + 0.0) / 45.0) * 100.0
 
 
-def get_pokemon_num(res):
+def extract_items_dict(res):
     inventory_delta = res['responses']['GET_INVENTORY'].get('inventory_delta', {})
     inventory_items = inventory_delta.get('inventory_items', [])
-    inventory_items_dict_list = map(lambda x: x.get('inventory_item_data', {}), inventory_items)
+    return map(lambda x: x.get('inventory_item_data', {}), inventory_items)
+
+
+def get_pokemon_num(res):
+    inventory_items_dict_list = extract_items_dict(res)
     inventory_items_pokemon_list = filter(lambda x: 'pokemon_data' in x and 'is_egg' not in x['pokemon_data'], inventory_items_dict_list)
     return len(inventory_items_pokemon_list)
 
 
 def get_inventory_data(res, poke_names):
-    inventory_delta = res['responses']['GET_INVENTORY'].get('inventory_delta', {})
-    inventory_items = inventory_delta.get('inventory_items', [])
-    inventory_items_dict_list = map(lambda x: x.get('inventory_item_data', {}), inventory_items)
+    inventory_items_dict_list = extract_items_dict(res)
     inventory_items_pokemon_list = filter(lambda x: 'pokemon_data' in x and 'is_egg' not in x['pokemon_data'],
                                           inventory_items_dict_list)
     inventory_items_pokemon_list_sorted = sorted(inventory_items_pokemon_list, key=lambda x: poke_names[str(x['pokemon_data']['pokemon_id'])].encode('ascii', 'ignore'))
@@ -40,11 +42,17 @@ def get_inventory_data(res, poke_names):
 
 
 def get_incubators_stat(res):
-    inventory_delta = res['responses']['GET_INVENTORY'].get('inventory_delta', {})
-    inventory_items = inventory_delta.get('inventory_items', [])
-    inventory_items_incubators = map(lambda x: x.get('inventory_item_data', {}), inventory_items)
-    inventory_items_dict_list = map(lambda x: x.get('egg_incubators', {}), inventory_items_incubators)
-    inventory_items_incubator_list = filter(lambda x: 'egg_incubator' in x and 'target_km_walked' in x, inventory_items_dict_list)
+    inventory_items_dict_list = extract_items_dict(res)
+    inventory_items_incubators = map(lambda x: x.get('egg_incubators', {}).get('egg_incubator', {}), inventory_items_dict_list)
+    inventory_items_incubator_list = reduce(lambda x, y: x + y, filter(lambda x: len(x) > 0, inventory_items_incubators))
+    player_stats = get_player_stats(res)
+    km_walked = player_stats.get('km_walked', 'None')
+    if inventory_items_incubator_list:
+        return (os.linesep.join(map(lambda x: "Incubator {0:.2f} km, walked {1:.2f} km".format(
+            x['target_km_walked'],
+            km_walked), inventory_items_incubator_list)))
+    else:
+        return 'No incubators'
     if inventory_items_incubator_list:
         return (os.linesep.join(map(lambda x: "Incubator {0:.2f} km, walked {1:.2f} km".format(
             x['egg_incubator']['target_km_walked'],

